File: src/Python/array/74.搜索二维矩阵.py
#
# @lc app=leetcode.cn id=74 lang=python
#
# [74] 搜索二维矩阵
#

# @lc code=start
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        left = 0
        right = len(matrix) * len(matrix[0]) - 1
        # binary search
        while left <= right:
            mid = (left + right) // 2
            # locate element in the matrix
            midVal = matrix[mid // len(matrix[0])][mid % len(matrix[0])]
            if midVal == target: return True
            elif midVal > target: right = mid - 1
            else : left = mid + 1
        return False
# @lc code=end

File: src/Java/array/74.搜索二维矩阵.java
/*
 * @lc app=leetcode.cn id=74 lang=java
 *
 * [74] 搜索二维矩阵
 */

// @lc code=start
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int left = 0, right = matrix.length * matrix[0].length - 1;
        // binary search
        while (left <= right) {
            int mid = (left + right) / 2;
            // loacte element in matrix
            int midVal = matrix[mid / matrix[0].length][mid % matrix[0].length];
            if (midVal == target) return true;
            else if (midVal > target) right = mid - 1;
            else left = mid + 1;
        }
        return false;
    }
}
// @lc code=end

